### PR TITLE
Add dynamic generative modules syntax (module system and GraphQL support)

### PR DESCRIPTION
**Pull request by antas-marcin**
### What's being changed:

This PR adds dynamic generative syntax. It adds a possibility of running multiple `generative modules` at one, without providing the needed configuration as with new, enhanced syntax we are able now to pass the settings dynamically.

This PR is a 1st PR that introduces changes in the module system and `GraphQL` syntax support, `gRPC` support will come later in the next PR.

This feature is fenced with `EXPERIMENTAL_DYNAMIC_RAG_SYNTAX` flag, if someone wants to give it a try then `EXPERIMENTAL_DYNAMIC_RAG_SYNTAX=true` environment variable needs to be passed to Weaviate.

All of the provider settings are now are grouped in it's own namespace, for example if one wants to pass settings for OpenAI now he can do this using this GraphQL syntax, example:

```
_additional{
  generate(
    singleResult:{
      prompt:"Create a short tweet about: {description}"
      openai:{
        model:"gpt-4"
        temperature:0.2
        n:10
      }
    }
  ){
    error
    singleResult
  }
}
```

It's also possible to return the usage statistics of a given provider and return debug information (such as generated prompt):

```
_additional{
  generate(
    singleResult:{
      prompt:"Create a short tweet about: {description}"
      debug:true
      google:{
        topP:0.5
        topK:2
      }
    }
  ){
    error
    singleResult
    debug{
      prompt
    }
    google{
      usageMetadata{
        candidatesTokenCount
        promptTokenCount
        totalTokenCount
      }
    }
  }
}
```

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: usecases/modules/modules.go [hunk 1]
```diff
@@ -306,6 +312,9 @@ func (p *Provider) shouldIncludeArgument(schema *models.Schema, module string,
 }
 
 func (p *Provider) shouldAddGenericArgument(class *models.Class, moduleType modulecapabilities.ModuleType) bool {
+	if moduleType == modulecapabilities.Text2TextGenerative {
+		return true
+	}
```
Comment (dirkkul) [inline usecases/modules/modules.go:317]: use p.isGenerativeModule() ?

File: usecases/modules/modules.go [hunk 2]
```diff
@@ -523,8 +548,21 @@ func (p *Provider) ExtractAdditionalField(className, name string, params []*ast.
 	if err != nil {
 		return err
 	}
+	additionalGenerativeDefaultProvider := ""
+	additionalGenerativeParameters := map[string]modulecapabilities.GenerativeProperty{}
 	for _, module := range p.GetAll() {
-		if p.shouldIncludeClassArgument(class, module.Name(), module.Type()) {
+		if name == "generate" {
```
Comment (dirkkul) [inline usecases/modules/modules.go:554]: can we use constants instead of strings?

**Review by dirkkul — COMMENTED** [review]
Could you add an e2e/acceptance test with the generative-dummy module?

There is already a test here, that can be easily extended: test/acceptance_with_go_client/generative_test.go and we could add one parameter  to the module to make it more "realistic"

File: modules/generative-aws/parameters/graphql.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+				},
+				"temperature": &graphql.InputObjectFieldConfig{
+					Description: "temperature",
+					Type:        graphql.Float,
+				},
+			},
+		}),
+		DefaultValue: nil,
+	}
+}
```
[the remark this answers is not in the document]
Comment (antas-marcin) [reply modules/generative-aws/parameters/graphql.go:38]: I was thinking about this but those really vary depending on the given API, the names are different for each API some can be unified such OpenAI API compliant ones (I was planning on doing in the next PR) but some not. Let's agree that I will tackle this in next PR.

File: modules/generative-aws/parameters/provider.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+import "github.com/weaviate/weaviate/entities/modulecapabilities"
+
+const name = "aws"
+
+func AdditionalGenerativeParameters(client modulecapabilities.GenerativeClient) map[string]modulecapabilities.GenerativeProperty {
+	return map[string]modulecapabilities.GenerativeProperty{
+		name: {Client: client, RequestParamsFunction: input, ResponseParamsFunction: nil, ExtractRequestParamsFunction: extract},
+	}
+}
```
[the remark this answers is not in the document]
Comment (antas-marcin) [reply modules/generative-aws/parameters/provider.go:22]: Let me do this in the next PR

File: modules/generative-cohere/clients/cohere.go [hunk 5]
```diff
@@ -163,14 +217,8 @@ func (v *cohere) generateForPrompt(textProperties map[string]string, prompt stri
...
-		}
-	}
-	// try getting header from GRPC if not successful
-	if apiKey := modulecomponents.GetValueFromGRPC(ctx, key); len(apiKey) > 0 && len(apiKey[0]) > 0 {
-		return apiKey[0]
+	if apiKey := modulecomponents.GetValueFromContext(ctx, key); apiKey != "" {
+		return apiKey
 	}
 	return ""
 }
```
[the remark this answers is not in the document]
Comment (antas-marcin) [reply modules/generative-cohere/clients/cohere.go:224]: Yeah, was doing that in all the places where I saw we can do it

**Comment by antas-marcin** [comment]
@dirkkul as it goes for dummy module I didn't add any additional input / output params there. I have just adjusted it to work with current changes. I will expand this module with input / output params and I will update your tests in the PR, ok?

File: modules/generative-aws/parameters/graphql.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dirkkul) [reply modules/generative-aws/parameters/graphql.go:38]: oki

**Review by dirkkul — APPROVED** [review]

**Comment by ccfontes** [comment]
Can we pass env vars like `EXPERIMENTAL_DYNAMIC_RAG_SYNTAX` in the cloud instance or is that just for self-hosted?

**Comment by dirkkul** [comment]
hey @ccfontes - that env var has been removed a while ago and dynamic RAG should work with all recent (1.30+) weaviate versions out of the box. In case it doesn't please ask in our community forum! :)

**Pull request merged**